src/test/ui/impl-trait/equality2.stderr

```diff
@@ -25,6 +31,12 @@ LL |     let _: i32 = Leak::leak(hide(0_i32));
 error[E0308]: mismatched types
   --> $DIR/equality2.rs:38:10
    |
+LL | fn hide<T: Foo>(x: T) -> impl Foo {
```

Comment:
I think it would be a good idea to add an additional explanation when the mismatch is due to the substs of am opaque type, as the current message seems quite self-contradictory. Maybe something like:

```
the expected opaque type (created from foo::<u32>)
the found opaque type (created from foo::<i32>)
note: a distinct opaque type is created for each distinct set of generic arguments provided to `foo`
```
`t

Reply:
I'm thinking about how to properly tackle this case. It is certainly confusing, but I would like to tackle it specifically and not as part of this PR. Would you mind filing a follow up ticket for this with your thoughts, @Aaron1011?

I think the wording you present is reasonable, but might require some tweaking, as the `created from foo::<_>` messages might be hard to accomplish.

Reply:
Opened https://github.com/rust-lang/rust/issues/68020